packages/blocks/src/store/selectors.js

```diff
@@ -95,6 +95,38 @@ export function getBlockVariations( state, blockName, scope ) {
 	} );
 }
 
+/**
+ * Returns the active block variation for a given block based on its attributes.
+ *
+ * @param {Object}                state      Data state.
+ * @param {string}                blockName  Name of block (example: “core/columns”).
+ * @param {Object}                attributes Block attributes used to determine active variation.
+ * @param {WPBlockVariationScope} [scope]    Block variation scope name.
+ *
+ * @return {(WPBlockVariation|void)} Active block variation.
+ */
+export function getActiveBlockVariation( state, blockName, attributes, scope ) {
+	const variations = getBlockVariations( state, blockName, scope );
+
+	const match = variations?.find( ( variation ) => {
+		if ( Array.isArray( variation.isActive ) ) {
+			const blockType = getBlockType( state, blockName );
+			const attributeKeys = Object.keys( blockType.attributes || {} );
+			return variation.isActive
+				.filter( ( attribute ) => attributeKeys.includes( attribute ) )
```

Comment:
Optional: Would it be worth validating this earlier during block registration? It might be nice for folks to know that something isn't quite right when developing blocks.

Reply:
Actually that's what I had in mind too in my first comment. This way we will avoid calling this check so many times.

Reply:
Good point! First iteration is here: https://github.com/WordPress/gutenberg/pull/30913/commits/637b62bbd79f78daeece9f1ca6c7f89a552a0cff